Docs: Mint/burn use case (#3250)
input-output-hk/cardano-wallet: docs/user-guide/common-use-cases/assets.md

Please note that you would mint and burn other assets pretty much the same way. In our example we will be additionally adding on-chain metadata to our minting transaction as we want our NFT to be CIP-25 compliant. However this step is not required for minting. For instance, you could mint some assets and add off-chain metadata for them ([CIP26](https://cips.cardano.org/cips/cip26)) in the [Cardano Token Registry](https://developers.cardano.org/docs/native-tokens/token-registry/cardano-token-registry).

### Minting an NFT

Let's see how can we mint an NFT with [CIP-25](https://cips.cardano.org/cips/cip25/) metadata using `cardano-wallet`.

#### Policy key

Before we attempt any minting/burning transaction we need to make sure our wallet is set up with a policy key. We can check it using [`GET /wallets/{walletId}/policy-key`](https://input-output-hk.github.io/cardano-wallet/api/edge/#operation/getPolicyKey):
```
$ curl -X GET http://localhost:8090/v2/wallets/73d38c71e4b8b5d71769622ab4f5bfdedbb7c39d/policy-key
"policy_vk12d0gdel9u6px8wf3uv4z6m4h447n9qsad24gztaku8dzzdqfajzqfm3rr0"
```
Looks good.

> In case we get `missing_policy_public_key` error:

Reviewer: ```suggestion
Otherwise we get `missing_policy_public_key` error:
```

```
$ curl -X GET http://localhost:8090/v2/wallets/73d38c71e4b8b5d71769622ab4f5bfdedbb7c39d/policy-key
{
  "code": "missing_policy_public_key",
  "message": "It seems the wallet lacks a policy public key. Therefore it's not possible to create a minting/burning transaction or get a policy id. Please first POST to endpoint /wallets/{walletId}/policy-key to set a policy key."
}

```

> We just need to make a `POST` request to [`POST /wallets/{walletId}/policy-key`](https://input-output-hk.github.io/cardano-wallet/api/edge/#operation/postPolicyKey) as suggested in the error message.

Reviewer: ```suggestion
In such a case we just need to make a `POST` request to [`POST /wallets/{walletId}/policy-key`](https://input-output-hk.github.io/cardano-wallet/api/edge/#operation/postPolicyKey) as suggested in the error message.
```

```
$ curl -X POST http://localhost:8091/v2/wallets/73d38c71e4b8b5d71769622ab4f5bfdedbb7c39d/policy-key \  
-d '{"passphrase":"Secure Passphrase"}' \  
-H "Content-Type: application/json"

"policy_vk12d0gdel9u6px8wf3uv4z6m4h447n9qsad24gztaku8dzzdqfajzqfm3rr0"
```

Once we sort it out we are all set and we can proceed into minting an NFT from our wallet!

#### CIP-25 metadata

We will be attaching CIP-25 metadata to our minting transaction, so there are few things that need to be sorted out. Most basic metadata can look as follows:
```
{
      "721": {
        "<POLICY_ID>": {
          "<ASSET_NAME>": {
            "name": "My amazing NFT",
            "image": "ipfs://<IPFS_ID>"
          }
        }
      }
}
```

As we can see we need to figure out `<POLICY_ID>`, `<ASSET_NAME>` and `<IPFS_ID>`.

##### Policy ID

> Policy id is basically a hash of the native script that guards minting operation. In case of Shelley wallets we can only sign with one key, a wallet spending key, but because of the fact that we can embed it into a native script we can practically have unlimited amount of policy ids from one wallet.

Reviewer: ```suggestion
Policy id is basically a hash of the native script that guards minting operation. In case of Shelley wallets we can only sign with one key, a wallet policy key, but because of the fact that we can embed it into a native script we can practically have unlimited amount of policy ids from one wallet.
```

These are examples of native scripts templates and each of them will produce different policy id.

```
cosigner#0

{ "all": [ "cosigner#0" ] }

{ "any": [ "cosigner#0" ] }

{ "some": {"at_least": 1, "from": [ "cosigner#0" ]} }

{ "all":
     [ "cosigner#0",
       { "active_from": 120 }
     ]
}

{ "all":
     [ "cosigner#0",
       { "active_until": 1200000 }
     ]
}
```

> > :information_source: `cosigner#0` stands for our wallet's spending key. In case of Shelley wallet we have only one. In the future, in the Shared wallets, we'll be able to have many spending keys shared between different users and they will be identified as `cosigner#1`, `cosigner#2`...

Reviewer: ```suggestion
> :information_source: `cosigner#0` stands for our wallet's policy key. In case of Shelley wallet we have only one. In the future, in the Shared wallets, we'll be able to construct a minting/burning script with many policy keys shared between different users and they will be identified as `cosigner#1`, `cosigner#2`...
```

{
  "policy_id": "6d5052088183db1ef06439a9f501b52721c2645532a50254a69d5390"
}
```

##### Asset name

The asset name acts as a sub-identifier within a given policy. Although we call it "asset name", the value needn't be text, and it could even be empty. In case of CIP-25 however we can use a plain text. Let our asset name be... `AmazingNFT`.

##### IPFS id

Let's assume that our NFT will point to a simple image, which we have already uploaded to IPFS. It is available at: https://ipfs.io/ipfs/QmRhTTbUrPYEw3mJGGhQqQST9k86v1DPBiTTWJGKDJsVFw. As we can see the IPFS id of the image is `QmRhTTbUrPYEw3mJGGhQqQST9k86v1DPBiTTWJGKDJsVFw`.

Now we can put together complete CIP-25 metadata JSON which we will use in the minting transaction:

```
{
      "721": {
        "6d5052088183db1ef06439a9f501b52721c2645532a50254a69d5390": {
          "AmazingNFT": {
            "name": "My amazing NFT",
            "image": "ipfs://QmRhTTbUrPYEw3mJGGhQqQST9k86v1DPBiTTWJGKDJsVFw"
          }
        }
      }
}
```
#### Minting transaction

We have already:
 - verified that our wallet is equipped with policy key
 - created CIP-25 metadata JSON.

We are now ready to mint!

> We will **construct**  transaction that mints 1 `AmazingNFT` with policy id derived from simple native script template = `cosigner#0` and posts related CIP-25 metadata to blockchain.

Reviewer: ```suggestion
We will **construct**  transaction that mints 1 `AmazingNFT` with policy id derived from simple native script template = `cosigner#0` and post the related CIP-25 metadata to blockchain.
```

Note that the wallet expects asset name to be hex-encoded string so let's hex encode our `AmazingNFT` first:
```
$ echo -n "AmazingNFT" | xxd -p
416d617a696e674e4654
```

Let's now [`POST /wallets/{walletId}/transactions-construct`](https://input-output-hk.github.io/cardano-wallet/api/edge/#operation/constructTransaction):

```
$ curl -X POST http://localhost:8090/v2/wallets/73d38c71e4b8b5d71769622ab4f5bfdedbb7c39d/transactions-construct \
-d '{
   "metadata":{
      "721":{
         "6d5052088183db1ef06439a9f501b52721c2645532a50254a69d5390":{
            "AmazingNFT":{
               "name":"My amazing NFT",
               "image":"ipfs://QmRhTTbUrPYEw3mJGGhQqQST9k86v1DPBiTTWJGKDJsVFw"
            }
         }
      }
   },
   "mint_burn":[
      {
         "operation":{
            "mint":{
               "quantity":1
            }
         },
         "policy_script_template":"cosigner#0",
         "asset_name":"416d617a696e674e4654"
      }
   ]
}' \
-H "Content-Type: application/json"
```
That's it! I should now receive CBOR-encoded `transaction`, `fee` and `coin_selection` details in response.

> I can now **sign** and **submit** such transaction just like in [[how-to-make-a-transaction]].

Reviewer: ```suggestion
I can now **sign** and **submit** such a transaction just like in [[how-to-make-a-transaction]].
```
